feat: allow conditional skip on success and fail comments

babblebey opened the pull request:
This PR Intend to transfer the flow from https://github.com/semantic-release/gitlab/pull/730 to the github plugin. It introduces the new properties into the configuration option, which are used to perform the stated function.

- `failCommentCondition` - used to state the conditions by which the `failComment` is posted or not
- `successCommentCondition` - used to state the conditions by which the `successComment` is posted or not and where it is posted

#### Changes Made

- Added deprecation notice for `failTitle`, `failComment` and `successComment` set as `false`
- Introduced logic to fetch more data for `issues` parsed from pull request body and commit messages to aid `successCommentCondition` filtering (more datails in https://github.com/semantic-release/github/pull/874#issuecomment-2294015820)
- Introduced utility to help perform stated functionality
  - `buildRelatedIssuesQuery` - Builds GraphQL query for fetching PRs/Commits related Issues to a list of commit hash (sha) from parsedIssue number
  - `buildAssociatedPRsQuery` - Builds GraphQL query for fetching associated PRs to a list of commit hash (sha)
  - `buildIssuesOrPRsFromResponseNode` - Build associatedPRs or RelatedIssues object (into issue-like object with `pull_request` property) from the GraphQL repository response


### Related Issue

Closes https://github.com/semantic-release/github/pull/360
Closes #359
Closes #666
Resolves #529

### Screencast/Screenshot

- **Demo `"successCommentCondition": false` - Do not add Success comment to any Issue or PR**

https://github.com/user-attachments/assets/63917a26-c71e-433b-9039-a667c37f43f1

- **Demo `"successCommentCondition": "<% return issue.pull_request; %>"` - Add Success comment to only Pull Requests**

https://github.com/user-attachments/assets/1f6b57fd-4b3b-4ca3-a4a6-cedde190f01c

- **Demo `"successCommentCondition": "<% return !issue.pull_request; %>"` - Add Success comment to only Issues**

https://github.com/user-attachments/assets/f392b1af-c2c6-4d9d-a4e3-dcd6f154ebf5

babblebey commented on lib/success.js:
```diff
@@ -56,6 +58,10 @@ export default async function success(pluginConfig, context, { Octokit }) {
 
   if (successComment === false) {
     logger.log("Skip commenting on issues and pull requests.");
+    logger.error(`Issue and pull request comments should be disabled via 'successCommentCondition'.
+      Using 'false' for 'successComment' is deprecated and will be removed in a future major version.`);
```
Hello @fgreinacher 👋, 

I am merely transferring this functionality/flow from the **GitLab** plugin to this place. But i'm a little bit lost on the deprecation talk of the `successComment and `failComment`, could you kindly shed some small light. 

🤔

fgreinacher replied:
@babblebey Sure, the relevant discussion is here: https://github.com/semantic-release/gitlab/pull/730#issuecomment-2180348168

We decided to deprecate the old way of disabling comments and gently push folks to migrate to the new condition options.

babblebey replied:
Thank you for the clarity @fgreinacher 💟

fgreinacher replied:
Sure thing @babblebey, cool that you're picking it up so quickly!

babblebey commented on test/fail.test.js:
```diff
@@ -447,3 +447,82 @@ test('Skip if "failTitle" is "false"', async (t) => {
+  const owner = "test_user";
+  const repo = "test_repo";
+  const env = { GITHUB_TOKEN: "github_token" };
+  const pluginConfig = { failCommentCondition: false };
```
Is `false` a valid value for `failCommentCondition` or is `"<% return false; %>"` the valid value otherwise? or Maybe Both??

Reason why I ask this is because the expression `"<% return false; %>"` doesn't resolve to `false` at some point.... like when....

fgreinacher replied:
Yes, `false` is a valid value as well.

We support that case so that we don't can save some API calls for that case. (Just an optimization). See original discussion at https://github.com/semantic-release/gitlab/pull/730#discussion_r1648070219.

babblebey replied:
So, I have a confirmation of my understanding now 😁, this means that it's a resolution that `false` causes the operation to stop early, while `"<% return false; %>"` stop the operation after making the the call to getting repo details 😉.

This I understood through the tests, and thanks for confirming the thought 🙏

babblebey commented on lib/fail.js:
```diff
@@ -23,14 +23,19 @@ export default async function fail(pluginConfig, context, { Octokit }) {
     logger.log("Skip issue creation.");
+    logger.error(`Failure reporting should be disabled via 'failCommentCondition'.
+      Using 'false' for 'failComment' or 'failTitle' is deprecated and will be removed in a future major version.`);
+  } else if (failCommentCondition === false) {
+    logger.log("Skip issue creation.");
   } else {
```
...when `failCommentCondition` is set to `"<% return false; %>"`, this block isn't executed, and I believe its because the template isn't resolved to `false` at that point yet..

babblebey reviewed (commented):
Hello @fgreinacher, 

Kindly help me clarify these observations... see review comments..

babblebey commented on lib/fail.js:
```diff
@@ -23,14 +23,19 @@ export default async function fail(pluginConfig, context, { Octokit }) {
     assignees,
   } = resolveConfig(pluginConfig, context);
 
   if (failComment === false || failTitle === false) {
```
Hello there @fgreinacher, @JonasSchubert 

On another note, deprecation of the `failTitle`, what does this mean for `failCommentCondition`??

Considering the `failTitle` is used as value for issue title in the api request to create an issue for failure, its absence is a problem when `failCommentCondition` is `true` and `failTitle` is `false`.

My point is `failCommentCondition` (in our current implementation) doesn't look like it can be used on it's own without having to enable both `failComment` and `failTitle` (kindly correct me if I perceive this wrongly). Considering the established implementation that uses `lodash.template`, are we looking at the possibility of incorporating the fail comment `title` and `body` inside `failCommentCondition` using templates??? and of course split'em back out later???

Is this the plan for that future MAJOR release? 🤔

fgreinacher replied:
No worries, the plan is only to remove support for disabling comments by setting successComment, failComment, or failTitle to false. It's totally fine to leave them undefined or set them to any other value.

JonasSchubert commented on test/fail.test.js:
```diff
@@ -447,3 +447,149 @@ test('Skip if "failTitle" is "false"', async (t) => {
+  t.true(t.context.log.calledWith("Skip commenting on or creating an issue."));
+});
+
+test(`Post new issue if none exists yet, but don't comment on existing issues when "failCommentCondition" is disallows it`, async (t) => {
```
```suggestion
test(`Post new issue if none exists yet, but don't comment on existing issues when "failCommentCondition" disallows it`, async (t) => {
```

JonasSchubert reviewed (changes requested):
As @fgreinacher said, awesome you picked that up so quickly, @babblebey , thanks a lot! Just found a small typo in one unit test description, beside that it looks good for me.
Will you also add unit tests for `success.js`?

babblebey commented:
> Will you also add unit tests for success.js?

Yea, working on that 😉

gr2m commented on lib/success.js:
```diff
@@ -56,6 +58,10 @@ export default async function success(pluginConfig, context, { Octokit }) {
 
   if (successComment === false) {
     logger.log("Skip commenting on issues and pull requests.");
+    logger.error(`Issue and pull request comments should be disabled via 'successCommentCondition'.
+      Using 'false' for 'successComment' is deprecated and will be removed in a future major version.`);
```
use `logger.warn` and prefix with `DEPRECATION:`, something like

> `Setting successComment to false is deprecated, please use successCommentCondition instead

```suggestion
    logger.warn(`DEPRECATION: 'false' for 'successComment' is deprecated and will be removed in a future major version. Use 'successCommentCondition' instead.`);
```

gr2m commented on lib/fail.js:
```diff
@@ -23,14 +23,19 @@ export default async function fail(pluginConfig, context, { Octokit }) {
   if (failComment === false || failTitle === false) {
     logger.log("Skip issue creation.");
+    logger.error(`Failure reporting should be disabled via 'failCommentCondition'.
+      Using 'false' for 'failComment' or 'failTitle' is deprecated and will be removed in a future major version.`);
```
same as my comment below

gr2m commented on lib/fail.js:
```diff
@@ -52,31 +57,45 @@ export default async function fail(pluginConfig, context, { Octokit }) {
+      ? template(failCommentCondition)({ ...context, issue: srIssue })
+      : true;
+
+    if (canCommentOnOrCreateIssue) {
```
exit early

I would do

```js
if (!canCommentOnOrCreateIssue) {
  logger.log("Skip commenting on or creating an issue.");
  return
}
```

https://www.google.com/search?q=return+early+pattern

JonasSchubert commented on lib/success.js:
```diff
@@ -130,6 +138,15 @@ export default async function success(pluginConfig, context, { Octokit }) {
+          ? template(successCommentCondition)({ ...context, issue })
+          : true;
+
+        if (!canCommentOnIssue) {
```
What do you think of applying `.filter` before `.map` to avoid the iteration with `undefined` values?

babblebey replied:
Sounds good to catch an early error cause.

but I assume that the `uniqBy()` operation already filters out any item that doesn't satisfy the conditions of the iteratee i.e. the value to specify uniqueness. 🤔

JonasSchubert commented on README.md:
```diff
@@ -80,24 +80,26 @@ When using the _GITHUB_TOKEN_, the **minimum required permissions** are:
+| `proxy`                   | The proxy to use to access the GitHub API. Set to `false` to disable usage of proxy. See [proxy](#proxy).                                                                                              | `HTTP_PROXY` environment variable.                                                                                                                   |
+| `assets`                  | An array of files to upload to the release. See [assets](#assets).                                                                                                                                     | -                                                                                                                                                    |
+| `successComment`          | The comment to add to each issue and pull request resolved by the release. Set to `false` to disable commenting on issues and pull requests. See [successComment](#successcomment).                    | `:tada: This issue has been resolved in version ${nextRelease.version} :tada:\n\nThe release is available on [GitHub release](<github_release_url>)` |
+| `successCommentCondition` | Use this as condition, when to comment on issues or merge requests. See [successCommentCondition](#successCommentCondition)                                                                            | -                                                                                                                                                    |
```
```suggestion
| `successCommentCondition` | Use this as condition, when to comment on issues or pull requests. See [successCommentCondition](#successCommentCondition)                                                                            | -                                                                                                                                                    |
```

Different phrasing for GitHub and Gitlab ;)

JonasSchubert commented on README.md:
```diff
@@ -199,6 +223,29 @@ The `failComment` `This release from branch ${branch.name} had failed due to the
+
+##### failCommentCondition example
+
+The `failComment` `This release from branch ${branch.name} had failed due to the following errors:\n- ${errors.map(err => err.message).join('\\n- ')}` will generate the comment:
```
I think this example belongs to `failComment`.

babblebey commented on README.md:
```diff
@@ -181,6 +183,28 @@ The `successComment` `This ${issue.pull_request ? 'pull request' : 'issue'} is i
+| `nextRelease` | `Object` with `version`, `channel`, `gitTag`, `gitHead` and `notes` of the release being done.                                                                                                                                                                                                               |
+| `commits`     | `Array` of commit `Object`s with `hash`, `subject`, `body` `message` and `author`.                                                                                                                                                                                                                           |
+| `releases`    | `Array` with a release `Object`s for each release published, with optional release data such as `name` and `url`.                                                                                                                                                                                            |
+| `issue`       | A [GitHub API Pull Request object](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request) for pull requests related to a commit, or an `Object` with the `number` property for issues resolved via [keywords](https://help.github.com/articles/closing-issues-using-keywords) |
```
# Todo 

- Justify this statement by building that `pull_request` object from the GraphQL associatedPRs response

babblebey commented on lib/success.js:
```diff
@@ -275,3 +290,25 @@ export function buildAssociatedPRsQuery(shas) {
     }
   `;
 }
+
+/**
+ * Build associatedPRs (to issue-like object with `pull_request` property) from the GraphQL repository response
+ * @param {object} repository
+ * @returns {object[]}
+ */
+function buildAssociatedPRs(repository) {
+  const associatedPRs = [];
+  for (const commit in repository) {
+    for (const node of repository[commit].associatedPullRequests.nodes) {
+      const pr = {
+        number: node.number,
+        body: node.body,
+        pull_request: {
+          url: node.url,
+        },
+      };
+      associatedPRs.push(pr);
+    }
+  }
+  return associatedPRs;
+}
```
- Utility to help build the PR object as `issue` value required for filtering in the `success|failCommentCondition`

babblebey reviewed (commented):
### Todo

...in review comments.

### Discovery

It looks like issue linked as `link` is kind of a problem really as reported here https://github.com/semantic-release/github/issues/586.... I've experienced this during my testing to adding successComment to only issue. Issues linked by link are not receiving the comment and that one is weird because in my initial tests without the filters at all, it was getting added..

I'll investigate further though 🤞

babblebey commented:
> I'll investigate further though 🤞

@gr2m turns out the reason why the issue isn't getting seen is because of the `owner/repo` details in the url not matching the current repo details.

Quick background: I changed the repo name from `sr` to `sr-github`.

So the linked issue url is `https://github.com/babblebey/sr/issues/11` instead of `https://github.com/babblebey/sr-github/issues/11` ....and that was getting seen as though it was an issue linked to an external repo which was filtered out in the logic here https://github.com/semantic-release/github/blob/51ee80881bdf1c1f2b52e8130f46c6158a114bad/lib/success.js#L117-L120

fgreinacher reviewed (approved):
Nice work 👏

gr2m commented on lib/success.js:
```diff
@@ -59,6 +61,12 @@ export default async function success(pluginConfig, context, { Octokit }) {
       logger.log("No commits found in release");
     }
     logger.log("Skip commenting on issues and pull requests.");
+    // TODO: use logger.warn() instead of logger.log()
```
There is no `logger.warn` right now, it would need to be added in `semantic-release` core

travi replied:
Just a quick fyi to keep in mind: to use the newly added logger in core, we'll need to bump the minimum bound of the peer dependency, which is a breaking change

babblebey replied:
Will do that in a follow-up PR 😉

gr2m reviewed (approved):
### 🚀

JonasSchubert commented:
Looking great!
<h6>and sorry for the late (or nearly no) response (moved to a new house without internet for 8 weeks...)</h6>

babblebey commented:
### New Commit

- Added new property `type` to the `issue.user` object... This was done to allow for more configuration ability.. hence patching a solution to https://github.com/semantic-release/github/issues/529 via configuration.
- Added test for cases where you don't want success Comment and labels added to PRs created by `Bot` e.g. dependabot etc.... this was easily done by configuring `successCommentCondition` as seen below

    ```js
    successCommentCondition: "<% return !issue.user || issue.user.type !== 'Bot'; %>"
    // This adds comment/label to all issues (regardless of user) and pull request created by users (ignoring bots)
    ```

😉

babblebey commented:
### Convert PR to Draft 

I have converted PR to draft for the following reason...

- Currently, when we fetch `associatedPRs` on commits, we get a sophisticated pull request object similar to that of the [get-a-pull-request ](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request) endpoint response object, which makes filtering more configurable in the `successCommentCondition` using properties like `labels`, `user` etc... BUT when we parse the body of the `associatedPRs` we get the related `issue`, but currently we only have an object with the `number` property for the issue its object, like so...

    ```js
    // issue
    {
      number: 23
    }
    
    // pull_request
    {
      pull_request: true,
      number: 45,
      title: "some title",
      body: "body markdown content",
      // ......more properties
    }
    ```
    meaning we do not have any means of performing any filtering similar to how we'd do on pull requests with the `successCommentCondition` in the case of related issues to associatedPRs.
    
    Rightfully, the properties present on the `associatedPRs` pull request object should also be present in the related issue object too to enable consistency in filtering operation at configuring `successCommentCondition`. 

### Way Forward

Currently, we simply parse the `associatedPRs` `body` to get the related issue number only.... Moving forward, I will implement a collection for all these parsed issue numbers and leverage the GitHub Issue GraphQL API, make a call using fragments to get details for these issues with their number and build that object similar to that of the pull request's.

😉
